File: Java/Permutations.java
public class Solution {
    
    private List<List<Integer>> ret;
    public List<List<Integer>> permute(int[] nums) {
        ret = new ArrayList<List<Integer>>();
        int[] index = new int[nums.length];
        for(int i=0; i<nums.length; i++) index[i] = -1;
        Arrays.sort(nums);
        backtrack(nums, 0, index);
        return ret;
    }
    
    private void backtrack(int[] nums, int k, int[] index) {
        if(k == nums.length) {
            Integer[] xx = new Integer[nums.length];
            for(int i=0; i<nums.length; i++) {
                xx[index[i]] = nums[i];
            }
            ret.add(Arrays.asList(xx));
            return;
        }
        
        for(int i=0; i<nums.length; i++) {
            // Make sure no duplicate permutation here after the check.
            // if(i>0 && nums[i] == nums[i-1] && index[i-1] == -1) continue;
            
            if(index[i] == -1) {
                index[i] = k;
                backtrack(nums, k+1, index);
                index[i] = -1;
            }
        }
    }
}

File: Swift/permutations.swift
/**
 * https://leetcode.com/problems/permutations/
 * 
 * 
 */ 
class Solution {
    func permute(_ nums: [Int]) -> [[Int]] {
        var nums = nums
        var ret: [[Int]] = []
        process(&ret, [], &nums)
        return ret
    }
    
    /// Try out all the possible choice for the next element in this `solution` (permutation)
    /// If we have none choice left in candidate, it means we have all choices in current `solution` (permutation), then it's a valid solution.
    ///
    ///
    /// - Parameters:
    ///     - result: final result, which contains all the solutions (possible permutations.)
    ///     - solution: current working partial solution, could be one of possible permutation.
    ///     - candidate: all the possible choices left to choose for the rest of current permutation.
    ///
    fileprivate func process(_ result: inout [[Int]], _ solution: [Int], _ candidate: inout [Int]) {
        if candidate.isEmpty {
            result.append(solution)
            return
        }
        for index in 0 ..< candidate.count {
            let tmp = candidate[index]
            candidate.remove(at: index)
            process(&result, solution + [tmp], &candidate)
            candidate.insert(tmp, at: index)
        }
    }
}
